pkg/schema/tracker.go

```diff
@@ -24,28 +24,67 @@ import (
 	"github.com/pingcap/parser/model"
 	"github.com/pingcap/tidb-tools/pkg/dbutil"
 	"github.com/pingcap/tidb-tools/pkg/filter"
+	tidbConfig "github.com/pingcap/tidb/config"
 	"github.com/pingcap/tidb/domain"
 	"github.com/pingcap/tidb/infoschema"
 	"github.com/pingcap/tidb/kv"
 	"github.com/pingcap/tidb/meta"
 	"github.com/pingcap/tidb/session"
 	"github.com/pingcap/tidb/store/mockstore"
+
+	"github.com/pingcap/dm/pkg/conn"
+	tcontext "github.com/pingcap/dm/pkg/context"
 )
 
 const (
 	waitDDLRetryCount = 10
 	schemaLeaseTime   = 10 * time.Millisecond
 )
 
+var (
+	sessionVars = []string{"sql_mode", "tidb_skip_utf8_check"}
+)
+
 // Tracker is used to track schema locally.
 type Tracker struct {
 	store kv.Storage
 	dom   *domain.Domain
 	se    session.Session
 }
 
-// NewTracker creates a new tracker.
-func NewTracker(sessionCfg map[string]string) (*Tracker, error) {
+// NewTracker creates a new tracker. `sessionCfg` will be set as tracker's session variables if specified, or retrieve
+// some variable from downstream TiDB using `tidbConn`.
+func NewTracker(sessionCfg map[string]string, tidbConn *conn.BaseConn) (*Tracker, error) {
+	// NOTE: tidb uses a **global** config so can't isolate tracker's config from each other. If that isolation is needed,
+	// we might SetGlobalConfig before every call to tracker, or use some patch like https://github.com/bouk/monkey
+	toSet := tidbConfig.NewConfig()
+	toSet.AlterPrimaryKey = true
+	tidbConfig.StoreGlobalConfig(toSet)
```

Comment:
Any test for it? rest LGTM

Reply:
in unit test 
https://github.com/pingcap/dm/pull/1032/files#diff-16690a9b535181de13c90f2f8ded85f9R144